compiler.py

```diff
@@ -0,0 +1,637 @@
+        #      function names as Python does, just as variables that point to an object that happens to be callable.
+        #      In some ways, this would be better than your unPythonic assumption that each function name can have only
+        #      one def in a program, though it would cost an instruction-cycle each time a def is "exectuted" which is
+        #      a pretty minor cost.  (Regardless, we probably should make an unPythonic assumption that a given
+        #      function-name will always accept the same arguments, though I suppose if we keep pushing positional args
+        #      on the stack to pass them, then we needn't even assume that!)
+
+        if node.name in self._functions or node.name == 'print':
+            raise CompilerError(ERR_REDEF, node)
+
+        self._in_def = node.name
+        reg_ret = f'{REG_RET_COUNTER_PREFIX}{len(self._functions)}'
+
+        args = node.args
+        if any((
+                args.posonlyargs,
+                args.vararg,
+                args.kwonlyargs,
+                args.kw_defaults,
+                args.kwarg,
+                args.defaults,
+        )):
+            raise CompilerError(ERR_INVALID_DEF, node)
+
+        # TODO  many of the above could/should be supported
+
+        # TODO it's better to put functions at the end and not have to skip them as code, but jumps need fixing
+        self._ins.append('jump {} always')
+
+        prologue = len(self._ins)  # line number where the function definition begins
+        self._functions[node.name] = Function(start=prologue, argc=len(args.args))
+
+        self._ins.append(f'read {reg_ret} cell1 {REG_STACK}')
+        for arg in args.args:
+            self._ins.append(f'op sub {REG_STACK} {REG_STACK} 1')
+            self._ins.append(f'read {arg.arg} cell1 {REG_STACK}')
+
+        # TODO argument passing would be more efficient if function-calling simply stored the arguments into the relevant
+        #      variables directly, rather than slowly piping them through a memory block, and this would also obviate
+        #      the need for an associated memory block.  (This would *not* however allow for recursion to retain
+        #      different namespaces for different levels, but to allow that we'd also need to store old values for the
+        #      namespace in the stack too, which would be really taxing for slow mindustry processors!)
+
+        for subnode in node.body:
+            self.visit(subnode)
+
+        epilogue = len(self._ins)
+        # TODO better way to patch this (ins should be concrete types, not immutable strings)
+        #      Telos:  I've caught myself having this sort of thought many times over the years, but these strings are
+        #              so short that creating new ones with heavily optimized methods like format just is better/faster
+        #              than any sort of mutable data-type we'd be likely to come up with!

```

Comment:
Speed isn't really a concern. mlog programs can't be very long, and sticking with format strings would currently be a micro-optimization, if optimization at all.